lib/plugins/aws/package/compile/events/s3/index.js

```diff
@@ -10,9 +10,32 @@ class AwsCompileS3Events {
     this.options = options;
     this.provider = this.serverless.getProvider('aws');
 
-    // TODO: Complete schema, see https://github.com/serverless/serverless/issues/8031
     this.serverless.configSchemaHandler.defineFunctionEvent('aws', 's3', {
-      anyOf: [{ type: 'string' }, { type: 'object' }],
+      anyOf: [
+        { $ref: '#/definitions/awsS3BucketName' },
+        {
+          type: 'object',
+          properties: {
+            bucket: { $ref: '#/definitions/awsS3BucketName' },
+            event: { type: 'string', pattern: '^s3:.+$' },
+            existing: { type: 'boolean' },
+            rules: {
+              type: 'array',
+              items: {
+                type: 'object',
+                properties: {
+                  prefix: { type: 'string' },
+                  suffix: { type: 'string' },
```

Comment:
If I see correctly in code, it's expected to be a `Name` and `Value` properties: https://github.com/serverless/serverless/blob/fc34140f4ec03958564a5868b339c40056f6b04e/lib/plugins/aws/package/compile/events/s3/index.js#L94-L103

Reply:
I beg to differ. The error message does say that, but the actual code in lines 101-103 expects an object, takes the first key-value pair from it and creates a rule object with `Name` and `Value` properties.

Reply:
Thanks for clarification, indeed it's not `{ Name: name, Value: value }` format that we expect from user.

Still if I see correctly, it's `{  [prefix|sufix]: value }`, while this schema expects something as `[{ prefix: value, suffix: value }, ...]`, so still doesn't appear correct

Reply:
I still differ :-)
Looking at the file in `master` in [lines 92-93](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/package/compile/events/s3/index.js#L92-L93) we have `rules.forEach` which implies `rules` is an array, and then the `isPlainObject` check which implies each item in that array should be an object.
Then in lines [101-102](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/package/compile/events/s3/index.js#L101-L102) we take the first key of each item with `Object.keys(rule)[0]` and the corresponding value.
In the [docs](https://github.com/serverless/serverless/blob/master/docs/providers/aws/events/s3.md#setting-filter-rules) it shows that indeed the value for rules is a list, and each item in the list is an object with a single key-value pair (one item has a key named `suffix` and the second item has a key named `prefix`):
```yaml
functions:
  users:
    handler: users.handler
    events:
      - s3:
          bucket: photos
          event: s3:ObjectCreated:*
          rules:
            - prefix: uploads/
            - suffix: .jpg
```
Theoretically, this yaml will also be valid:
```yaml
functions:
  users:
    handler: users.handler
    events:
      - s3:
          bucket: photos
          event: s3:ObjectCreated:*
          rules:
            - prefix: uploads/
               foo: bar
            - suffix: .jpg
               baz: qux
```
But the end results will be the same because we take only the first key from each object-item in `rules`(`foo: bar` and `baz: qux` will be ignored).

Reply:
ugh.. indeed 🤦 Sorry for confusion. It's pretty quirky construct after all.